fix: improve transformation of compliant username

MatousJobanek opened the pull request:
improvement of the transformation of the compliant username:
* drop `*@domain.com` suffix from the username. The reason is that a label in a domain name may contain 63 chars max. We need to keep in mind that the username is used also for the names of the namespaces and the namespace name is part of routes created in it. We should avoid creating unnecessarily long usernames otherwise someone could run out of chars per label while creating routes.
* if there is any special character, then we replace it by a hyphen `-`
* if there are more hyphens in a sequence, then we replace it by the single one
* if there is a name that starts with a hyphen, then we add a prefix `crt-`
* if there is a name that ends with a hyphen, then we add a suffix `-crt`
* if there is a name that contains only numbers, then we add a prefix `crt-`

paired PR: https://github.com/codeready-toolchain/toolchain-e2e/pull/94

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller.go:
```diff
@@ -390,6 +395,24 @@ func (r *ReconcileUserSignup) generateCompliantUsername(instance *toolchainv1alp
 }
 
+func transformUsername(username string) string {
+	userName := specialCharRegexp.ReplaceAllString(strings.Split(username, "@")[0], "-")
```
It's actually possible to create a user with username starting with `@`. I've just checked it. In this case the username will be an empty string if I'm not missing anything. Let's cover this case too.
For example we could transform `@spiderman` to `crt-spiderman` or to `at-spiderman`. I like `at-spiderman` better but `crt-spiderman` would work too.

xcoulon replied:
WWIW, I prefer the `at-` prefix (or `-at` suffix)

MatousJobanek replied:
Thanks for pointing this out - I missed this case.
done: 5b4422c

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1866,3 +1846,33 @@ func newObjectMeta(name, email string) metav1.ObjectMeta {
+	assertName(t, "some", "some@email.com")
+	assertName(t, "so-me", "so-me@email.com")
+	assertName(t, "some", "some")
+	assertName(t, "so-me", "so-me")
```
```suggestion
	assertName(t, "so-me", "so-me")
	assertName(t, "so-me", "so-----me")
```

MatousJobanek replied:
this is already indirectly covered by `assertName(t, "so-me", "so:#$%!$%^&me")` but it's good to add also the direct test. Thanks :+1:

xcoulon commented:
> improvement of the transformation of the compliant username:
> 
> * drop `*@domain.com` suffix from the username. The reason is that a label in a domain name may contain 63 chars max. We need to keep in mind that the username is used also for the names of the namespaces and the namespace name is part of routes created in it. We should avoid creating unnecessarily long usernames otherwise someone could run out of chars per label while creating routes.

I may be missing something, but is it possible that 2 users have similar usernames, one as the form of an email address (eg: `johndoe@domain.com`) and one as just a regular name (eg: `johndoe`). In that case, will be have a collision, or will we end with something like `johndoe` and `johndoe-2`?

alexeykazakov commented:
We will end up with johndoe-2 (if it’s not taken) in this case. Or johndoe-3
if *-2 is taken and so on.

On Mon, Apr 6, 2020 at 12:52 AM Xavier Coulon <notifications@github.com>
wrote:

> improvement of the transformation of the compliant username:
>
>    - drop *@domain.com suffix from the username. The reason is that a
>    label in a domain name may contain 63 chars max. We need to keep in mind
>    that the username is used also for the names of the namespaces and the
>    namespace name is part of routes created in it. We should avoid creating
>    unnecessarily long usernames otherwise someone could run out of chars per
>    label while creating routes.
>
> I may be missing something, but is it possible that 2 users have similar
> usernames, one as the form of an email address (eg: johndoe@domain.com)
> and one as just a regular name (eg: johndoe). In that case, will be have
> a collision, or will we end with something like johndoe and johndoe-2?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/codeready-toolchain/host-operator/pull/176#issuecomment-609634046>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEXMNYMXNHQKCWDJV64JILRLGC3XANCNFSM4L3627TQ>
> .
>

xcoulon commented:
> We will end up with johndoe-2 (if it’s not taken) in this case. Or johndoe-3 if *-2 is taken and so on.
> […](#)
> 

Ok, so that logic of avoiding the collisions is still in place and is not impacted by this change. That's what I wanted to confirm. Thanks!

MatousJobanek commented:
> Ok, so that logic of avoiding the collisions is still in place and is not impacted by this change. That's what I wanted to confirm. Thanks!

yes, that logic is not impacted and it should be also covered by this test: https://github.com/codeready-toolchain/host-operator/blob/d142a45476f4712dd22b003df1b03f6beb8db841/pkg/controller/usersignup/usersignup_controller_test.go#L901-L964
and here you can see that the logic will try 100 numbers and then it will fail: https://github.com/codeready-toolchain/host-operator/blob/d142a45476f4712dd22b003df1b03f6beb8db841/pkg/controller/usersignup/usersignup_controller_test.go#L1368-L1430

MatousJobanek commented:
> We will end up with johndoe-2 (if it’s not taken) in this case. Or johndoe-3
if *-2 is taken and so on.

it actually starts appending the numbers with `-1`, which is probably not intuitive/obvious - I'll change it to `-2`

MatousJobanek commented:
> it actually starts appending the numbers with -1, which is probably not intuitive/obvious - I'll change it to -2

changed in 032ad1d and modified e2e tests in: https://github.com/codeready-toolchain/toolchain-e2e/pull/94/commits/ccb941d87833d378f3a60ff75c29677f971f5959
